### PR TITLE
Add support for CNMeM library.

### PR DESCRIPTION
**Pull request by jdemouth**
Update the code to support CNMeM (formerly known as Cumem).

### CODE REVIEW

File: theano/sandbox/cuda/cuda_ndarray.cu [hunk 1]
```diff
@@ -111,7 +111,6 @@ int initCumem(int card_number_provided, int card_nb) {
         ///@TODO: thejaswi: add support for multiple streams
         devices[i].numStreams = 0;
         devices[i].streams = NULL;
-        devices[i].granularity = 0;
```
Comment (nouiz) [inline theano/sandbox/cuda/cuda_ndarray.cu]: I think you need to set streamSizes to NULL to be safe. Idem bellow. But maybe your code don't read it when streams is NULL, but I prefer to be save and don't have not initialized ptr around.

File: theano/sandbox/cuda/cuda_ndarray.cu [hunk 2]
```diff
@@ -152,12 +150,15 @@ void * device_malloc(size_t size, int verbose)
     #endif
     void * rval=NULL;
     ///@TODO: thejaswi: support for multiple-streams?
-    if(g_use_cumem) {
-        cumemStatus_t status = cumemMalloc(&rval, size, NULL);
-        if(status != CUMEM_STATUS_SUCCESS) {
+    if(g_use_cnmem) {
+        cnmemStatus_t status = CNMEM_STATUS_SUCCESS;
+        if( size != 0 ) {
+            status = cnmemMalloc(&rval, size, NULL);
```
Comment (nouiz) [inline theano/sandbox/cuda/cuda_ndarray.cu]: For my knowledge, why don't do the call when size is 0? Do the lib work in that case and this is just a small speed up?

**Comment by nouiz** [comment]
Except for the small comment and questions, it is good. thanks.

**Comment by jdemouth** [comment]
Hi Frederic. I have addressed your comments. I have also updated CNMEM because we found a bug in the library. You should be good to go.

**Comment by nouiz** [comment]
thanks!

**Pull request merged**